### PR TITLE
at86rf2xx: do not set src pan compression on init

### PR DESCRIPTION
**Pull request by cgundogan**
Depends on ~~#5525~~ #5685 

Currently, the at86rf2xx driver sets the src pan compression on initialization. Doing this when building the 802.15.4 header instead (like e.g. proposed by @aeneby in ~~#5525~~ #5685 ) is more dynamic, so that the flag is still set/unset correctly *iff* the pan id changes.

### CODE REVIEW

**Comment by miri64** [comment]
Do you mean #5685? Also: What happens if the flag *is* set?

**Comment by cgundogan** [comment]
yes #5685 

>Also: What happens if the flag is set?

Can you rephrase? The flag *is* set by default currently and this leads to the src pan being compressed (elided from the header). This is no problem for now (and correct), because there is no case where `ieee802154_set_Frame_hdr()` is called with different src and dst pans (https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/link_layer/netdev2/gnrc_netdev2_ieee802154.c#L185).

*iff* `ieee802154_set_Frame_hdr()` gets called with different pan ids in the future, then this flag must be unset (this is still missing in #5685 )

**Comment by miri64** [comment]
You unset it, yes. But what happens now, if someone sets it later on?

**Comment by cgundogan** [comment]
>You unset it, yes. But what happens now, if someone sets it later on?

If you set it later, then you have the same functionality as before.

**Comment by cgundogan** [comment]
as a side note, `NETDEV2_IEEE802154_PAN_COMP` does not correspond to a register of the device.

**Comment by aeneby** [comment]
I don't really understand how PAN compression is relevant to the driver at all... is the `NETDEV2_IEEE802154_PAN_COMP` flag even necessary?

**Comment by miri64** [comment]
> I don't really understand how PAN compression is relevant to the driver at all... is the NETDEV2_IEEE802154_PAN_COMP flag even necessary?

If src_pan == dst_pan but the compression shouldn't be used eg.  I already had the problem, that there is a bug in lwIP that it just ignores that bit and always skips the source PAN.

**Comment by aeneby** [comment]
> If src_pan == dst_pan but the compression shouldn't be used

That would be an invalid frame though, according to the 802.15.4 spec.

**Comment by miri64** [comment]
Yes, but for Interop its sometimes necessary to provide behavior beyond the spec until a bug is fixed ;-) (I don't say that its not wrong, just that lwIP is doing it wrong atm).

**Comment by aeneby** [comment]
I don't really know enough about lwIP to comment, but if lwIP is doing it wrong then it needs to be fixed in lwIP, not broken in RIOT - two wrongs don't make a right :)

**Comment by thomaseichinger** [comment]
>>If src_pan == dst_pan but the compression shouldn't be used

>That would be an invalid frame though, according to the 802.15.4 spec.

@aeneby Could you point me to where in the 802.15.4 spec this is specified? Or which version of 802.15.4? As far as I read point 5.2.1.1.5 it is possible to have the two PAN IDs set to the same value without the need to use compression.

**Comment by aeneby** [comment]
@thomaseichinger, you have a point. In the 2003 spec it was quite clear, since it was called the "Intra-PAN" field. With this definition, it simply doesn't make sense to have src and dst PANs the same without this bit being set (and therefore the src PAN removed). However in the 2011 spec, the name has changed to "PAN ID Compression" field, even though the functional definition remains basically the same. I guess this would allow Intra-PAN packets to be sent with both src and dst PANs specified, according to the semantic definition of this field...?

Now I'm not 100% sure, I will think over it.

**Comment by miri64** [comment]
@aeneby 

> I don't really know enough about lwIP to comment, but if lwIP is doing it wrong then it needs to be fixed in lwIP, not broken in RIOT - two wrongs don't make a right :)

I agree with you and I finally came around to [open an issue](https://savannah.nongnu.org/bugs/index.php?48825) on the lwIP side, but I know enough about lwIP to say that changes to mainline are quite slow ([6lowpan took 1.5 years to get to master](http://git.savannah.gnu.org/cgit/lwip.git/log/?h=lowpan6), my issue about the incorrect fragmentation [a very pressing issue] is [still open after ~6 month](https://savannah.nongnu.org/bugs/?47291) and the last non-RC release was [nearly 4 years ago](http://git.savannah.gnu.org/cgit/lwip.git/commit/?h=STABLE-1_4_1&id=5b8b5d459e7dd890724515bbfad86c705234f9ec)). So as a workaround I would still have this kind of option provided (in case they provide the fragmentation fix first ^^).

> @thomaseichinger, you have a point. In the 2003 spec it was quite clear, since it was called the "Intra-PAN" field. With this definition, it simply doesn't make sense to have src and dst PANs the same without this bit being set (and therefore the src PAN removed). However in the 2011 spec, the name has changed to "PAN ID Compression" field, even though the functional definition remains basically the same. I guess this would allow Intra-PAN packets to be sent with both src and dst PANs specified, according to the semantic definition of this field...?
>
> Now I'm not 100% sure, I will think over it.

To me it reads that the compression is basically optional. Yes it makes sense to compress the source PAN away if it is identical, but I wouldn't say its completely wrong to include it regardless.

**Comment by miri64** [comment]
(to be clear: the faulty behavior of lwIP in my eyes is that they completely ignore the compression bit, not that they include the source PAN, even if it is identical to the destination PAN).

**Comment by thomaseichinger** [comment]
@miri64 This is undisputed and the discussion not necessarily prevents from moving forward with this PR as AFAIK RIOT should be able to deal with two same PAN IDs anyways.

**Comment by miri64** [comment]
> @miri64 This is undisputed and the discussion not necessarily prevents from moving forward with this PR as AFAIK RIOT should be able to deal with two same PAN IDs anyways.

But #5685 takes away this option.

**Comment by thomaseichinger** [comment]
@miri64 I think it's totally standard conform if RIOT does PAN ID compression, my note was more about the receiving side. RIOT should be able to process a frame that doesn't have the compression bit set and two identical PAN IDs.

**Comment by aeneby** [comment]
After doing some careful reading, here is my interpretation (I'll try to stay relevant to this PR):

- According to the 802.15.4-2003 spec, _any frame with identical source and destination PANs is invalid_
- In the 802.15.4-2011 spec, the meaning of the field has changed; _it is valid, albeit wasteful, for a frame to contain identical source and destination PANs_
- In the 802.15.4-2015 spec... who knows, because I don't enjoy reading standards documents enough to pay ~$400 for the privilege. Does anyone have access to this?

What does this mean for this PR?

- It does in fact make some sense to have the `NETDEV2_IEEE802154_PAN_COMP` flag since, in theory at least, it's valid to send intra-PAN packets without using PAN compression
- In practice though, I don't see why you would ever want to do this. Doesn't that mean this should be left as `true` by default?
- While #5685 does enforce the only really sensible behaviour, strictly speaking it is perhaps wrong to do so - we should not be modifying this flag based on src and dst PANs, but rather using this flag to _determine what to do if the src and dst PANs are the same_ (compress or not compress)

The main issue is that, since each version of the 802.15.4 supersedes the ones before it, 802.15.4-2015 should be the one we are following. So there is still a missing piece of the puzzle here.

**Comment by aeneby** [comment]
> RIOT should be able to process a frame that doesn't have the compression bit set and two identical PAN IDs.

According to my interpretation of the 2011 spec, yes, this is correct. Interesting to note though that Linux will drop such frames.

**Comment by aeneby** [comment]
See https://github.com/RIOT-OS/RIOT/pull/5685#issuecomment-241309161.

So in short, it looks like frames with identical src and dst PAN IDs are invalid again in the 2015 spec. Further, it's clear that the MAC layer should be automatically handling PAN ID compression. So actually, provided the linked text is accurate, I don't see any reason for the `NETDEV2_IEEE802154_PAN_COMP` flag to exist at all.

File: drivers/at86rf2xx/at86rf2xx.c [hunk 1]
```diff
@@ -100,7 +100,7 @@ void at86rf2xx_reset(at86rf2xx_t *dev)
     /* set default TX power */
     at86rf2xx_set_txpower(dev, AT86RF2XX_DEFAULT_TXPOWER);
     /* set default options */
-    at86rf2xx_set_option(dev, NETDEV2_IEEE802154_PAN_COMP, true);
+    at86rf2xx_set_option(dev, NETDEV2_IEEE802154_PAN_COMP, false);
```
Comment (miri64) [inline drivers/at86rf2xx/at86rf2xx.c]: If this option is ignored anyway, shouldn't the code for `at86rf2xx_set_option` be disabled anyway?

Comment (miri64) [inline drivers/at86rf2xx/at86rf2xx.c]: For this option I mean.

Comment (PeterKietzmann) [inline drivers/at86rf2xx/at86rf2xx.c]: @miri64 I think you're right. There is no explicit implementation behind this option, so simply removing this line would probably be the most logical thing?

**Comment by PeterKietzmann** [comment]
How about me use this PR just for removing `at86rf2xx_set_option(dev, NETDEV2_IEEE802154_PAN_COMP, false);` from the driver and handle the rest in #5685

**Comment by miri64** [comment]
@PeterKietzmann +1

File: sys/net/link_layer/ieee802154/ieee802154.c [hunk 2]
```diff
@@ -27,6 +27,10 @@ size_t ieee802154_set_frame_hdr(uint8_t *buf, const uint8_t *src, size_t src_len
     uint8_t type = (flags & IEEE802154_FCF_TYPE_MASK);
     uint8_t bcast = (flags & IEEE802154_BCAST);
 
+    if (dst_pan.u16 == src_pan.u16) {
+        flags |= IEEE802154_FCF_PAN_COMP;
+    }
```
Comment (miri64) [inline sys/net/link_layer/ieee802154/ieee802154.c]: See https://github.com/RIOT-OS/RIOT/pull/5759#issuecomment-244012587

**Comment by miri64** [comment]
Alternative in https://github.com/RIOT-OS/RIOT/pull/5897

**Comment by cgundogan** [comment]
No objection to close once #5897 is merged.

**Comment by PeterKietzmann** [comment]
Closing in favour of #5897.
